### PR TITLE
Get Icon URL Request

### PR DESCRIPTION
**Pull request by JackLivio**
### Risk
This PR makes **minor** API changes.

### Testing Plan
Add icon_url to policy table app_policies.
Register app. 
Verify onSystemRequest sent to mobile, and SystemRequest received from mobile.
AppIcons for cloud app should appear in the storage folder.

### Summary
Add new RequestType to be used by onSystemRequest and SystemRequest.
After the first app registers, for each cloud app with an icon_url in the policy table that do not have an app icon yet, core will send an onSystemRequest to mobile. 

Mobile will retrieve the icon from the url and send the binary data back to core via a SystemRequest. The fileName will be the image url used in the onSystemRequest. Core will save this icon in the app icon folder under the name of the cloud apps policy id.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/src/application_manager_impl.cc [hunk 1]
```diff
@@ -3703,6 +3799,66 @@ void ApplicationManagerImpl::SendDriverDistractionState(
...
 }
 
+void ApplicationManagerImpl::SendGetIconUrlNotifications(
+    const uint32_t connection_key, ApplicationSharedPtr application) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  std::vector<std::string> enabled_apps;
+  GetPolicyHandler().GetEnabledCloudApps(enabled_apps);
+  std::vector<std::string>::iterator enabled_it = enabled_apps.begin();
+  std::vector<std::string>::iterator enabled_end = enabled_apps.end();
+  app_icon_map_lock_ptr_.Acquire();
```
Comment (jacobkeeler) [inline src/components/application_manager/src/application_manager_impl.cc]: AutoLock could be used here

File: src/components/application_manager/src/message_helper/message_helper.cc [hunk 2]
```diff
@@ -1547,6 +1547,8 @@ bool MessageHelper::CreateHMIApplicationStruct(
   message[hmi_response::policy_app_id] = policy_app_id;
 
   const std::string icon_path = app->app_icon_path();
+
+  LOG4CXX_DEBUG(logger_, "Get ICON PATH" << icon_path);
```
Comment (jacobkeeler) [inline src/components/application_manager/src/message_helper/message_helper.cc]: Looks like a testing log, should probably just remove

File: src/components/include/policy/policy_external/policy/policy_manager.h [hunk 3]
```diff
@@ -114,6 +114,15 @@ class PolicyManager : public usage_statistics::StatisticsManager {
    */
   virtual std::string GetLockScreenIconUrl() const = 0;
 
+  /**
+   * @brief Get Icon Url used for showing a cloud apps icon before the intial
+   *registration
+   *
+   * @return url which point to the resourse where icon could be
```
Comment (jacobkeeler) [inline src/components/include/policy/policy_external/policy/policy_manager.h]: *`resource`

File: src/components/include/policy/policy_regular/policy/policy_manager.h [hunk 4]
```diff
@@ -92,6 +92,15 @@ class PolicyManager : public usage_statistics::StatisticsManager {
    */
   virtual std::string GetLockScreenIconUrl() const = 0;
 
+  /**
+   * @brief Get Icon Url used for showing a cloud apps icon before the intial
+   *registration
+   *
+   * @return url which point to the resourse where icon could be
```
Comment (jacobkeeler) [inline src/components/include/policy/policy_regular/policy/policy_manager.h]: *`resource`

File: src/components/include/policy/policy_external/policy/policy_manager.h [hunk 5]
```diff
@@ -114,6 +114,15 @@ class PolicyManager : public usage_statistics::StatisticsManager {
    */
   virtual std::string GetLockScreenIconUrl() const = 0;
 
+  /**
+   * @brief Get Icon Url used for showing a cloud apps icon before the intial
```
Comment (jacobkeeler) [inline src/components/include/policy/policy_external/policy/policy_manager.h]: *`initial`

File: src/components/include/policy/policy_regular/policy/policy_manager.h [hunk 6]
```diff
@@ -92,6 +92,15 @@ class PolicyManager : public usage_statistics::StatisticsManager {
    */
   virtual std::string GetLockScreenIconUrl() const = 0;
 
+  /**
+   * @brief Get Icon Url used for showing a cloud apps icon before the intial
```
Comment (jacobkeeler) [inline src/components/include/policy/policy_regular/policy/policy_manager.h]: *`initial`

File: src/components/interfaces/MOBILE_API.xml [hunk 7]
```diff
@@ -2419,6 +2419,7 @@
         <element name="MEDIA" />
         <element name="FOTA" />
         <element name="OEM_SPECIFIC" since="5.0" />
+        <element name ="ICON_URL" />
```
Comment (jacobkeeler) [inline src/components/interfaces/MOBILE_API.xml]: remove stray space

Comment (jacobkeeler) [reply src/components/interfaces/MOBILE_API.xml]: `since="5.1"`

File: src/components/policy/policy_external/src/sql_pt_representation.cc [hunk 8]
```diff
@@ -1070,6 +1071,9 @@ bool SQLPTRepresentation::SaveSpecificAppPolicy(
   app.second.cloud_transport_type.is_initialized()
       ? app_query.Bind(10, *app.second.cloud_transport_type)
       : app_query.Bind(10);
+  app.second.icon_url.is_initialized()
+      ? app_query.Bind(10, *app.second.icon_url)
+      : app_query.Bind(10);
```
Comment (jacobkeeler) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: *`Bind(11)`

File: src/components/policy/policy_regular/src/cache_manager.cc [hunk 9]
```diff
@@ -892,6 +892,20 @@ std::string CacheManager::GetLockScreenIconUrl() const {
   return std::string("");
 }
 
+std::string CacheManager::GetIconUrl(const std::string& policy_app_id) const {
+  std::string url = "";
```
Comment (jacobkeeler) [inline src/components/policy/policy_regular/src/cache_manager.cc]: `CACHE_MANAGER_CHECK(std::string())`

File: src/components/policy/policy_external/src/cache_manager.cc [hunk 10]
```diff
@@ -1603,6 +1603,20 @@ std::string CacheManager::GetLockScreenIconUrl() const {
   return std::string("");
 }
 
+std::string CacheManager::GetIconUrl(const std::string& policy_app_id) const {
+  std::string url = "";
```
Comment (jacobkeeler) [inline src/components/policy/policy_external/src/cache_manager.cc]: `CACHE_MANAGER_CHECK(std::string())`

File: src/components/application_manager/src/application_manager_impl.cc [hunk 11]
```diff
@@ -3703,6 +3799,66 @@ void ApplicationManagerImpl::SendDriverDistractionState(
...
+    (*message)[strings::params][strings::connection_key] = connection_key;
+    (*message)[strings::params][strings::message_type] =
+        mobile_apis::messageType::notification;
+    (*message)[strings::params][strings::protocol_version] =
+        application->protocol_version();
+    (*message)[strings::msg_params][strings::request_type] =
+        mobile_apis::RequestType::ICON_URL;
+    (*message)[strings::msg_params][strings::url] = url;
+
+    // todo update to false when this request comes back
```
Comment (jacobkeeler) [inline src/components/application_manager/src/application_manager_impl.cc]: remove todo

File: src/components/application_manager/src/application_manager_impl.cc [hunk 12]
```diff
@@ -808,6 +822,51 @@ void ApplicationManagerImpl::OnHMIStartedCooperation() {
...
+  app_icon_map_lock_ptr_.Release();
+  return std::string("");
+}
+
+void ApplicationManagerImpl::SetIconExists(const std::string policy_id) {
+  app_icon_map_lock_ptr_.Acquire();
+  auto app_icon_it = app_icon_map_.find(policy_id);
+  if (app_icon_it != app_icon_map_.end()) {
+    LOG4CXX_DEBUG(logger_, "Updaing icon exists status");
+    app_icon_it->second.icon_exists = true;
```
Comment (jacobkeeler) [inline src/components/application_manager/src/application_manager_impl.cc]: Part of me thinks that the entry should just be erased from the map if the icon is present (and never added to the map in the first place if it was already present on start)

File: src/components/policy/policy_regular/src/cache_manager.cc (hunk 9 above)
Comment (jacobkeeler) [reply src/components/policy/policy_regular/src/cache_manager.cc]: Also, you could remove the `= ""`

File: src/components/application_manager/src/application_manager_impl.cc [hunk 13]
```diff
@@ -808,6 +821,52 @@ void ApplicationManagerImpl::OnHMIStartedCooperation() {
   RefreshCloudAppInformation();
 }
 
+std::string ApplicationManagerImpl::PolicyIDByIconUrl(const std::string url) {
+  app_icon_map_lock_ptr_.Acquire();
+  for (auto& x : app_icon_map_) {
+    auto policy_id = x.first;
+    std::string icon_url = GetPolicyHandler().GetIconUrl(policy_id);
+    if (icon_url == url) {
+      LOG4CXX_DEBUG(logger_, "Matched icon url: " << url);
```
[the remark this answers is not in the document]
Comment (JackLivio) [reply src/components/application_manager/src/application_manager_impl.cc]: Fixed. The code below was supposed to be in the if block.

File: src/components/application_manager/src/application_manager_impl.cc [hunk 14]
```diff
@@ -860,7 +859,17 @@ void ApplicationManagerImpl::SetIconExists(const std::string policy_id) {
     file.is_persistent = true;
     file.is_download_complete = true;
     file.file_name = full_icon_path;
-    file.file_type = mobile_apis::FileType::GRAPHIC_PNG;
+
+    std::string icon_url = GetPolicyHandler().GetIconUrl(policy_id);
+    std::string extension = boost::filesystem::extension(icon_url);
+    if (extension == "bmp" || extension == "BMP") {
+      file.file_type = mobile_apis::FileType::GRAPHIC_BMP;
+    } else if (extension == "JPEG" || extension == "jpeg") {
```
Comment (jacobkeeler) [inline src/components/application_manager/src/application_manager_impl.cc]: doesn't `.jpg` work here too?

**Pull request merged**